Creating a new User

ghost opened the pull request:
After a lot of trying things that didn't work, creating a user with or without an avatar is now functional, the create user page is mounted in place of the UserProfile fragment. You won't see a change on screen after pressing create, but the user will be created in the DB under All Accounts.  All fields must contain at least one character.

This answers user stories 1.02.01, 1.03.01 and 1.03.03, and partially 1.07.01 

Things that need to be done: Tests, there's also some weird bug causing the toolbar to float a little bit

<img width="1051" alt="Screenshot 2024-11-02 at 4 16 32 PM" src="https://github.com/user-attachments/assets/6b98a41a-c6ed-4c92-92fc-1d15ec181eb9">
<img width="1019" alt="Screenshot 2024-10-31 at 6 15 44 PM" src="https://github.com/user-attachments/assets/951e359e-ad1a-4745-86de-fe75c24d9f99">

Everything has been updated... just assigning the user stories to the branch now

SuperCreativeUniqueUsername commented on app/src/main/java/com/example/pygmyhippo/common/Account.java:
do you think it could make more sense to make the account connect to database in a new database connector class? I think keeping data structure classes and database classes separate makes more sense, but let me know if there was a particular reason you decided to do it this way.

Jrmckay2 replied:
Maybe, I wrote the original logic before the DB connector was bought up as a thing. I'm mostly just not sure what adapting that would look like?

SuperCreativeUniqueUsername replied:
I'm just thinking adding this kind of functionality to account doesn't really make sense. Maybe some other can weigh in on this @katzhong  @Kori-Kozicki @Jameszf @avreet-14

Jrmckay2 replied:
I changed it

SuperCreativeUniqueUsername commented on app/src/main/java/com/example/pygmyhippo/common/UserProcessing/UserProcessing.java:
```diff
@@ -0,0 +1,212 @@
+     * @version 1.0
+     * */
+    private void uploadAvatar(Bitmap bitmap) throws FileNotFoundException {
+        // Convert bitmap to byte array for upload
```
Are we going to be using bitmaps for image uploads? I am open to this idea as long as we all use bitmaps for images. I guess just make sure everyone knows to use bitmaps instead of firebase storage.

Jrmckay2 replied:
The user selected avatar uploads from a string, but the generated Avatar uploads from a byte array, the bitmap is just one step of the process. It's not ideal, but it's been extremely difficult to get generated avatars to upload at all, even moving the byteArray out of userProcessing will break it unless you move all the logic used to download the avatar as well.

SuperCreativeUniqueUsername replied:
I have no problem with it, it's just something that we should all be doing the same. So moving forward let's say all images are stored in bitmaps in the database.

Jrmckay2 replied:
I haven't had any luck with converting regular images into bitmaps, even small ones, it creates a null exception pointer. The image path is correct, as it's what is used to upload via string, I've tried adding a time delay in case it takes more time to process which hasn't worked either.

Kori-Kozicki reviewed (commented):
My initial comment is after rereading the user story, has phone number been made an optional field? Or will it break if we try not to enter it

Kori-Kozicki commented on app/src/main/java/com/example/pygmyhippo/common/Account.java:
```diff
@@ -83,25 +116,99 @@ public Account(
+            newUser.put("current_role", "entrant");
+            newUser.put("roles", userRoles);
+            AccountRef.document(deviceID).set(newUser);
+
```
I think we don't have to get over complicated here. I believe you can just add Java objects straight into the collection without having to make a whole new map. That way to when we retrieve the data it can come back as objects. This works but in the future we should probably consider that

Kori-Kozicki commented on app/src/main/java/com/example/pygmyhippo/common/UserProcessing/UserProcessing.java:
```diff
@@ -0,0 +1,212 @@
+        if (imagePath == null) {
+            this.avatarPath = imageUrl + userName;
+            loadAndUpload();
+            System.out.println("generate avatar..");
```
Is this System.out for debugging? If not what is it there for?

Jrmckay2 replied:
Just debugging

Jameszf replied:
Using Log.d or Log.i instead of System.out may be better

Kori-Kozicki commented on gradle/libs.versions.toml:
```diff
@@ -30,6 +32,8 @@ navigation-fragment = { group = "androidx.navigation", name = "navigation-fragme
 firebase-database = { group = "com.google.firebase", name = "firebase-database", version.ref = "firebaseDatabase" }
+firebase-storage = { group = "com.google.firebase", name = "firebase-storage", version.ref = "firebaseStorage" }
+firebase-installations = { group = "com.google.firebase", name = "firebase-installations", version.ref = "firebaseInstallations" }
 
```
Just a general question so I don't get lost: What does Firebase storage and installations do?

Jrmckay2 replied:
Storage is for picture, installations is for the Firebase Installation ID, which is what we are using for the device Id

ghost commented:
> My initial comment is after rereading the user story, has phone number been made an optional field? Or will it break if we try not to enter it

It won't break it just wouldn't submit if you enter nothing in the field, I can change that though

ghost commented:
> My initial comment is after rereading the user story, has phone number been made an optional field? Or will it break if we try not to enter it

I updated it, also added snack bars (alternative to toasts) which give popup if the user doesn’t enter a Name or email. Pronouns and phone number are optional. A snack bar also pops up to confirm a user was created.

Kori-Kozicki commented:
> > My initial comment is after rereading the user story, has phone number been made an optional field? Or will it break if we try not to enter it
> 
> I updated it, also added snack bars (alternative to toasts) which give popup if the user doesn’t enter a Name or email. Pronouns and phone number are optional. A snack bar also pops up to confirm a user was created.

That sounds good to me!

Kori-Kozicki reviewed (approved):
I think this mostly looks good to me. We should probably consider adding tests to this before it gets merged to main tho. Lately we've been lacking at that

Jameszf commented:
I will be talking over this PR

Jameszf commented:
This will be closed until I get a chance to understand the code.
